src/IO/ReadBufferFromS3.h

```diff
@@ -58,7 +63,7 @@ class ReadBufferFromS3 : public SeekableReadBufferWithSize
 
     void setReadUntilPosition(size_t position) override;
 
-    Range getRemainingReadRange() const override { return Range{ .left = static_cast<size_t>(offset), .right = read_until_position }; }
+    Range getRemainingReadRange() const override { return Range{.left = static_cast<size_t>(offset), .right = read_until_position - 1}; }
```

Comment:
I think it could be better to make `read_until_position` as optional here in this buffer. Because now it is 0 by default and if it was not updated then this `- 1` can lead to something unexpected.

Reply:
`read_until_position` is exclusive bound, so `0` already denotes no limit.
But this part is not safe, I'll return `std::nullopt` for `.right` if `read_until_position = 0`.